### ISSUE TITLE
DefaultedOnCreate object Is the default value instead of Salesforce value

### ISSUE THREAD

**Issue by noamby** [django-salesforce/django-salesforce#280]
Hi :)

I have an issue when creating a model with field using DefaultedOnCreate()
The field is not updated with the actual value in Salesforce, it's value would be DefaultedOnCreate object

Example Model:
```
class Contact(SalesforceModel):
  currency_iso_code = TestableCharField(db_column='CurrencyIsoCode', max_length=3, verbose_name='Contact Currency', default=models.DefaultedOnCreate('USD'), choices=[('USD', 'U.S. Dollar')], blank=True, null=True)
  last_name = models.CharField(max_length=80)
```

Example Code:
```
contact = Contact.objects.create(last_name="test contact")
contact.currency_iso_code
# will print <salesforce.backend.operations.DefaultedOnCreate object at 0x10f543cf8>
```

**Comment by hynekcer** [comment]
Yes, it is currently an expected feature of Salesforce that can be circumvented in several possible ways by you.  A simple information about it should be added to README. Too complicated information can be found in our [wiki about default=models.DEFAULTED_ON_CREATE](https://github.com/django-salesforce/django-salesforce/wiki/Introspection-and-Special-Attributes-of-Fields) A "fix" is also possible, but it would require a discussion, because a fix could be undesirable for some people. I explain the current state first.

Salesforce cloud has a feature that a very sophisticated method can be be used to determine a default value if a field is omitted in a new object that is created. Your default currency can be e.g. determined by settings of the organization and you can use the same code for many organizations that have different default currencies. If your case is not so complicated then you can simply replace it by `default='USD'`. A typical example and maybe the only example where a DefaultedOnCreate is necessary almost for everyone is the "Owner" field of the object that is can be a current user in a simplest possible case but not `None`.

Another solution is to refresh the value from the database: `contact = Contact.objects.get(pk=contact.pk)` or some [refresh function](https://github.com/django-salesforce/django-salesforce/blob/79ba20d6b03a88e72581e71917a3cc636ac21391/tests/test_mixin/tests.py#L6-L11) if you really need a more dynamic default assigned by Salesforce, not by Django.

A parameter in `models.DefaultedOnCreate('USD')` is intended only for possible tests of your application with SalesforceModel with other databases that have no DefauledOnCreate feature and it is ignored with the Salesforce database.

(Recently it is also possible to "fix" that issue in this framework by using a composite request instead of simple REST request to get also a new value of your field by same request that created the object.  
I will continue later, why it could be not a good solution...)

**Comment by hynekcer** [comment]
(... continue)
The model created by inspectdb should work easily. That's a right requirement. How we can achieve it:

1) Automatically refresh new values of fields with DefaultedOnCreate
2) Replace the field with DefaultedOnCreate in a new created object by a simple value, e.g. "USD in this case"
3) Use DefaultedOnCreate as rarely as possible

The solution 1 could cause new issues and can be a problematic in new releases.

I think that the solution 2 is more unacceptable than a current "bug". Imagine the current `default=models.DefaultedOnCreate('USD')`, but a formula is assigned in Salesforce that assigns "EUR". The contact is created in Salesforce with a value EUR and a guessed value USD is assigned by Django and the Django object is saved again.

I think that it is more important to preserve a normal unmodified save() method and it is less important to use DefaultedOnCreate so frequently as now by default, especially for people who don't need it. It can be configurable by a parameter in inspectdb to create models.py more compatible with Django or models more compatible with Salesforce.

**Comment by noamby** [comment]
@hynekcer thanks for the detailed explanation, for now I'm using `refresh_from_db()` on my models to get the updated value from SFDC
I have nothing to add since you wrote everything so detailed :)

**Comment by hynekcer** [comment]
This issue is fixed by a commit above and it will be used very soon in the version 4.1: A simple default value is used by `inspectdb` instead of `DefaultedByCreate(some_value)`.

All constant default values can be introspected. If a default value is a more complicated or unknown function is used by Salesforce cloud then still a symbolic default value `DEFAULTED_ON_CREATE` must be used to delegate the final assignment to Salesforce.

**Comment by hynekcer** [comment]
It is fixed for Django 5.0 by commits [c61f66b](https://github.com/django-salesforce/django-salesforce/commit/c61f66bf0180789d9ce6291222bb4891a3fb8c99) and [84e0186](https://github.com/django-salesforce/django-salesforce/commit/84e0186f779d27d5abbe432775f991c41015d42d).  See the last comment in #308

I added the promissed docs: [Introspection and Special Attributes of Fields](https://github.com/django-salesforce/django-salesforce/wiki/Introspection-and-Special-Attributes-of-Fields)
